File: 03-09-simple-thread/Main.py
import time
import threading
import random

class Echo(threading.Thread):
    def __init__(self, m):
        threading.Thread.__init__(self)
        self.m = m
        self.r = random.Random()

    def run(self):
        for i in range(20):
            self.randSleep()
            print "%d: %s" % (i, self.m)
        print "[%s] HAS WON...." % self.m

    def randSleep(self):
        time.sleep(self.r.randint(0, 1000) / 1000.0)

t1 = Echo("CSCI 2020U")
t2 = Echo("Sys Dev & Int")

t1.start()
t2.start()

t1.join();
t2.join();
print "========== ALL DONE ============"

File: 03-09-simple-thread/Main.java
import java.util.Random;

class Main {
    public static void main(String[] args) {
        Echo e1 = new Echo("CSCI 2020U");
        Echo e2 = new Echo("Sys Dev & Int");

        e1.start();
        e2.start();

        try {
            e1.join();
            e2.join();
            System.out.println("====== all done ========");
        } catch(InterruptedException e) {
            ;
        }
    }
}
class Echo extends Thread {
    String m;
    Random r = new Random();
    public Echo(String m) {
        this.m = m;
    }
    private void randSleep() {
        long t = r.nextInt(1000);
        try {
            Thread.sleep(t);
        } catch(InterruptedException e) {
            ;
        }
    }
    public void run() {
        for(int i=0; i < 20; i++) {
            this.randSleep();
            System.out.println(i + ": " + this.m);
        }
        System.out.println("[" + m + "] HAS WON......");
    }
}
